### DISCUSSION TITLE
CLI features to better analyze your app

### DISCUSSION THREAD

**Discussion by stephanrotolante** [vercel/next.js#62567] [Ideas]
### Goals

- Better visiblity on what pages are affected by file changes

### Background

I have worked on NextJS apps where QA teams are wasting valuable time doing full regression test for flows that are not affected by the new feature work. This time could be better spent focused on testing the flows were directly affected by the newly introduced code. With that knowledge teams could stream line their whole UAT process. Potentially change the way they ship code.

Using Typescript or JSDoc are great ways to migrate risk when ensuring no build errors; however this isn't a 100% guarantee that your app will work at runtime, as these static checks can be fooled + values aren't guaranteed in JS land.

Another way to migrate risk is a proper automated testing strategy; whether that is unit, integration, or e2e flows. Not all devs team can implement that or have a mature enough flow for it

I am not sure if there already exists a tool to do a file dependency check of each nextjs page, but I built a little poc.

### Proposal

I wrote a little POC of doing the dependency check:
https://github.com/stephanrotolante/scout

It would be nice to integrate something this into the nextjs cli. I think dev teams would greatly benefit

**Comment by icyJoseph** [comment]
Hi,

You can use https://www.npmjs.com/package/dependency-cruiser to map out and even apply rules to how imports/exports should behave. There's also ESLINT rules for that, https://eslint.org/docs/latest/rules/no-restricted-imports.

I think the dependency-cruiser people also made this tool, https://www.npmjs.com/package/watskeburt which can show, changed stuff, and direct and indirectly affected modules:

- https://github.com/sverweij/dependency-cruiser/issues/816#issuecomment-1597650269

**Reply by stephanrotolante** [reply]
@icyJoseph Thanks for the suggestions, I will check those packages out!

**Comment by icyJoseph** [comment]
I checked your project out and I think it's awesome 😎 ! Good stuff!

**Reply by stephanrotolante** [reply]
Thanks! I have been having a lot of fun with Go lately. I think the tool does a decent job of going through the files. The JS import system can get a little tricky at times
